### PR TITLE
Documentation: Document how the event loop works & what it does

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
This was long overdue; as of writing this I saw another question about something that seems obvious to me who spent a bunch of time with the system.

Hopefully this is generic enough so that we can change things without this getting out of date.

### CODE REVIEW

File: Documentation/EventLoop.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+-   Timer events, i.e. events that fire after a certain timeout, possibly repeatedly, can be created with `EventLoop::register_timer` and `Object::start_timer()`. A more user-friendly version is the `Core::Timer` utility class which does the same thing and allows you to attach any callback to the timer.
+-   For when a "file" becomes readable or writeable, the utility class `Core::Notifier` interfaces with the event loop system to handle exactly that.
+
+Note that all events are registered, and therefore happen, on the event loop of the thread that created the event.
+
+## Dos and Don'ts with event loops
+
+-   DO NOT store an event loop in a global variable. Because the event loop itself relies on global variable initialization, UBSAN will catch an initialization order fiasco. DO create your main event loop in `main` and pass it to classes which need it.
+-   DO NOT access the current event loop if you don't know on which thread you are running. If there is no event loop on your thread, the program will crash. DO recieve the specific event loop you need to talk to as an initializer variable.
```
Comment (GMTA) [inline Documentation/EventLoop.md:40]:
```suggestion
-   DO NOT access the current event loop if you don't know on which thread you are running. If there is no event loop on your thread, the program will crash. DO receive the specific event loop you need to talk to as an initializer variable.
```

**Comment by GMTA** [comment]
You mention the event loop stack a couple of times, but not why it exists. Maybe give an example of a common use case of the event loop stack?

File: Documentation/EventLoop.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+All of this applies per thread. The event loops share some global state (notifiers, timers, event loop stack etc.) which are thread-local variables. The convenience getter `EventLoop::current()` relies on this to give you the topmost event loop on the thread you're calling from.
+
+The **event loop stack** is mainly used for nesting GUI windows. Each window adds another event loop onto the stack, and remaining events are added to the lower event loop's queue when it exist. The purpose behind this system is that `GUI::Window`s and other systems which provide their own event loop can just create and run this class depending on global variables without interfering with other event loops.
```
Comment (GMTA) [inline Documentation/EventLoop.md:25]:
```suggestion
The **event loop stack** is mainly used for nesting GUI windows. Each window adds another event loop onto the stack, and remaining events are added to the lower event loop's queue when it exits. The purpose behind this system is that `GUI::Window`s and other systems which provide their own event loop can just create and run this class depending on global variables without interfering with other event loops.
```

**Review by GMTA — CHANGES_REQUESTED** [review]
LGTM, just a small typo!

File: Documentation/EventLoop.md [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+## What it can do
+
+An event loop handles several kinds of events:
+
+-   POSIX signals can be registered with `EventLoop::register_signal()`. This means that the event loop of the calling thread registers the specified POSIX signal and callback with the kernel, and you can be sure that the signal handler will run as a normal event without the weirdness that comes with POSIX signal handlers (such as unspecified thread).
+-   To fire an event on a `Core::Object` the next time the event loop runs, `EventLoop::post_event()` exists.
```
Comment (ADKaster) [inline Documentation/EventLoop.md:32]: Inverting the subject/object here feels more natural

> `EventLoop::post_event()` allows calling code to fire an event targeting a specific `Core::Object` the next time the event loop is pumped.

**Pull request merged**